src/v3/pact.ts

```diff
@@ -142,11 +152,113 @@ export class PactV3 {
           return Promise.reject(new Error(error))
         })
         .finally(() => {
-          this.pact.shutdownTest(result)
+          this.pact.shutdownMockServer(result.mockServer)
         })
     } else {
-      this.pact.shutdownTest(result)
+      this.pact.shutdownMockServer(result.mockServer)
       return Promise.reject(result.testError)
     }
   }
 }
+
+interface RunningServer {
+  pact: PactV3
+  mockServer: V3MockServer
+}
+
+interface TestResult {
+  mockServerError: string | null
+  mockServerMismatches: string[] | null
+}
+
+export async function withMockServer<T>(
+  testFn: (...mockServers: V3MockServer[]) => Promise<T>,
+  ...pacts: PactV3[]
+): Promise<T> {
+  const runningServers: RunningServer[] = []
+
+  try {
+    pacts.forEach(pact => {
+      runningServers.push({
+        pact,
+        mockServer: (pact as any).pact.startMockServer() as V3MockServer,
+      })
+    })
+
+    const value = await testFn(
+      ...runningServers.map(({ mockServer }) => mockServer)
+    )
+
+    const pactErrors = getTestResults(runningServers).filter(isError)
+
+    if (pactErrors.length) {
+      let message = "Mock server failed with the following mismatches: "
+      for (const pactError of pactErrors) {
+        message += formatPactErrorMessage(pactError)
+      }
+      throw new Error(message)
+    }
+
+    runningServers.forEach(({ pact, mockServer }) => {
+      const nativePact = (pact as any).pact
+      nativePact.writePactFile(mockServer.id, (pact as any).opts.dir)
+    })
+
+    return value
+  } catch (err) {
+    const pactErrors = getTestResults(runningServers).filter(isError)
+
+    if (pactErrors.length) {
+      let message = "Test failed for the following reasons:"
+      message += "\n\n\tTest code failed with an error: " + err.message
+
+      for (const pactError of pactErrors) {
+        message += formatPactErrorMessage(pactError)
+      }
+
+      const newError = new Error(message)
+      // 'Forward' original stack trace:
+      newError.stack = err.stack
+      throw newError
+    }
+
+    throw err
+  } finally {
+    runningServers.forEach(({ pact, mockServer }) => {
+      ;(pact as any).pact.shutdownMockServer(mockServer)
```

Comment:
No idea what's going on with this weird opening semicolon. Somehow `prettier` insists on placing it there.

Reply:
It may be the linter or prettier, I'll need to look at the code. It might be a case it's concerned there is ambiguity and has inserted a semi to prevent any compiler mistakes (the compiler can't always correctly infer the correct automatic semicolon insertion).

In this case, it does look wrong - don't worry for now!

I'll also look at this PR tomorrow with fresh eyes.